File: src/epay_tool/ebay/catalog.py
"""eBay Catalog API - identify a product from its barcode (GTIN/EAN/UPC).

This is the primary enrichment path: for most retail products a single
call returns the exact catalog entry with title, brand, aspects and
stock images - no AI involved.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from epay_tool.ebay.client import EbayClient


@dataclass
class CatalogMatch:
    epid: str
    title: str
    brand: str | None = None
    mpn: str | None = None
    image_urls: list[str] = field(default_factory=list)
    aspects: dict[str, list[str]] = field(default_factory=dict)
    category_ids: list[str] = field(default_factory=list)


def _first_mpn(value: str | list[str] | None) -> str | None:
    """The catalog returns the manufacturer part number as a string or a list."""
    if isinstance(value, list):
        return value[0] if value else None
    return value


def find_by_gtin(client: EbayClient, gtin: str) -> CatalogMatch | None:
    """Search the eBay catalog by barcode. Returns the best match or None."""
    data = client.get(
        "/commerce/catalog/v1/product_summary/search",
        params={"gtin": gtin, "limit": 5},
        token="app",
    )
    summaries = data.get("productSummaries") or []
    if not summaries:
        return None
    best = summaries[0]
    product_id = best.get("epid") or best.get("productId", "")
    match = CatalogMatch(
        epid=product_id,
        title=(best.get("title") or "")[:80],
        brand=best.get("brand"),
        mpn=_first_mpn(best.get("mpn")),
        image_urls=[
            img["imageUrl"]
            for img in ([best.get("image")] if best.get("image") else [])
            + (best.get("additionalImages") or [])
            if img and img.get("imageUrl")
        ],
        category_ids=best.get("categoryIds") or [],
    )
    for aspect in best.get("aspects") or []:
        name, values = aspect.get("localizedName"), aspect.get("localizedValues")
        if name and values:
            match.aspects[name] = values
    return match

File: src/main/ebay/catalog.ts
/**
 * eBay Catalog API - identify a product from its barcode (GTIN/EAN/UPC).
 *
 * This is the primary enrichment path: for most retail products a single
 * call returns the exact catalog entry with title, brand, aspects and
 * stock images - no AI involved.
 */

import type { EbayClient } from './client'

export const MAX_TITLE_LENGTH = 80

export interface CatalogMatch {
  epid: string
  title: string
  brand: string | null
  mpn: string | null
  imageUrls: string[]
  aspects: Record<string, string[]>
  categoryIds: string[]
}

interface ProductSummary {
  epid?: string
  productId?: string
  title?: string
  brand?: string
  mpn?: string | string[]
  image?: { imageUrl?: string }
  additionalImages?: Array<{ imageUrl?: string }>
  categoryIds?: string[]
  aspects?: Array<{ localizedName?: string; localizedValues?: string[] }>
}

/** The catalog returns the manufacturer part number as a string or a list. */
function firstMpn(value: string | string[] | undefined): string | null {
  if (Array.isArray(value)) return value[0] ?? null
  return value ?? null
}

/** Search the eBay catalog by barcode. Returns the best match or null. */
export async function findByGtin(client: EbayClient, gtin: string): Promise<CatalogMatch | null> {
  const data = await client.get<{ productSummaries?: ProductSummary[] }>(
    '/commerce/catalog/v1/product_summary/search',
    { params: { gtin, limit: 5 }, token: 'app' },
  )

  const best = data.productSummaries?.[0]
  if (!best) return null

  const aspects: Record<string, string[]> = {}
  for (const aspect of best.aspects ?? []) {
    if (aspect.localizedName && aspect.localizedValues?.length) {
      aspects[aspect.localizedName] = aspect.localizedValues
    }
  }

  const imageUrls = [best.image, ...(best.additionalImages ?? [])]
    .map((image) => image?.imageUrl)
    .filter((url): url is string => Boolean(url))

  return {
    epid: best.epid ?? best.productId ?? '',
    title: (best.title ?? '').slice(0, MAX_TITLE_LENGTH),
    brand: best.brand ?? null,
    mpn: firstMpn(best.mpn),
    imageUrls,
    aspects,
    categoryIds: best.categoryIds ?? [],
  }
}
